splunk_topology/check.py

```diff
@@ -105,7 +116,31 @@ def check(self, instance):
             self.start_snapshot(instance_key)
 
         try:
-            self._auth_session(instance)
+            if 'token' not in authentication:
+                self.log.debug("Using basic authentication mechanism")
+                self._auth_session(instance)
+            else:
+                self.log.debug("Using token based authentication mechanism")
+                persist_token_key = instance.instance_config.base_url + "token"
+                if self.status.data.get(persist_token_key) is None:
+                    self.log.debug("Creating a new token from first initial token")
+                    # Since this is first time when check starts with initial token,
+                    # we need to validate and create the new token
+                    if self._is_valid_token(instance, authentication["token"])[0]:
+                        new_token = self._create_auth_token(instance, authentication["token"])
+                        self.update_token_memory(instance.instance_config.base_url, new_token)
+                    else:
+                        msg = "Initial Token is expired, Please renew your token or use the valid token."
+                        raise Exception(msg)
+                token = self.status.data.get(persist_token_key)
+                token_flag, days_expiry = self._is_valid_token(instance, token)
+                if not token_flag:
+                    if days_expiry < 0:
+                        msg = "Token is expired, Please renew your token or use the valid token"
+                        raise Exception(msg)
+                    else:
+                        new_token = self._create_auth_token(instance, token)
```

Comment:
Some duplicate code/logic in this code block.

We're first checking if the memory contains a token, if not we create one. After that we check if the created or existing token is valid. If not, create a new one. Seems like some redundant logic.
Logic could be? pseudo:

```
token = get token from memory
if token is None:
  token = token from YAML
if self.is_expired(token):
  raise Exception("Current in use authentication token is expired. Please provide a valid token in the YAML and restart the Agent."

# check if we need to renew the token before expiry
if need_renewal(token)
  new_token = request_new_token()
  memory.update(new_token)
```

Let's keep the `is_valid` method pure. Logic; just check if the token is expired or not (maybe change the method's name to `is_expired`) and have another method to check whether the token is almost expired such that you can renew the token. This also helps in getting rid of the less beautiful tuple return type of `is_valid_token`

Reply:
@whazenberg both method `need_renewal` and `is_expired` will have the same implementation except returning `days` and `boolean` flag. So thought to put in the same otherwise I had the same logic implemented initially.